semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/linking/calllinker/HeuristicsCallLinker.scala

```diff
@@ -0,0 +1,102 @@
+import io.shiftleft.codepropertygraph.generated.{DispatchTypes, EdgeTypes}
+import io.shiftleft.passes.{CpgPass, DiffGraph}
+import io.shiftleft.semanticcpg.language._
+
+import scala.collection.mutable
+
+/*
+ * This pass uses simple heuristics to link CALL nodes to METHOD nodes.
+ * It is most useful for CPGs with incomplete type information, for example
+ * generated from the source of Kotlin programs.
+ *
+ * In this pass, we assume that the METHOD_FULL_NAME properties of both CALL
+ * and METHOD nodes have the format `$method_name:$method_ret_type($method_param_1_type, $method_param2_type, ...)`
+ *
+ * The heuristics used are:
+ * 1. If the CPG contains CALLs with a method name matching a single METHOD node,
+ * then add a CALL edge between them.
+ *
+ * 2. If the CPG contains CALLs with a method name matching multiple METHOD nodes with varying numbers of parameters,
+ * then add a CALL edge to the METHOD node with a number of parameters matching the number of arguments.
+ *
+ * 3. If the CPG contains CALLs with a method name matching multiple METHOD nodes with the same number of parameters,
+ * then add CALL edges to _all_ the METHOD nodes with a number of parameters matching the number of arguments.
+ */
+class HeuristicsCallLinker(cpg: Cpg) extends CpgPass(cpg) {
+
+  private val methodNameToNodes = mutable.Map.empty[String, Seq[StoredNode]]
+
+  override def run(): Iterator[DiffGraph] = {
+    val dstGraph = DiffGraph.newBuilder
+    cpg.method.foreach { method =>
+      val entry = methodNameToNodes.get(method.name)
+      if (entry.isDefined) {
+        methodNameToNodes.put(method.name, entry.get ++ Seq(method))
+      } else {
+        methodNameToNodes.put(method.name, Seq(method))
+      }
+    }
+
+    cpg.call.foreach { call =>
+      try {
+        linkCall(call, dstGraph)
+      } catch {
+        case exception: Exception =>
+          throw new RuntimeException(exception)
+      }
+    }
+
+    Iterator(dstGraph.build())
+  }
+
+  private def linkCall(call: Call, dstGraph: DiffGraph.Builder): Unit = {
+    call.dispatchType match {
+      case DispatchTypes.STATIC_DISPATCH =>
+        linkStaticCall(call, dstGraph)
+      case DispatchTypes.DYNAMIC_DISPATCH =>
+      // TODO: use logger
+      case _ => println(s"WARN: Unknown dispatch type on dynamic CALL ${call.code}")
+    }
+  }
+
+  private def linkStaticCall(call: Call, dstGraph: DiffGraph.Builder): Unit = {
+    val methodName = call.methodFullName.split(":").head
+    val resolvedMethodOption = methodNameToNodes.get(methodName)
+    if (resolvedMethodOption.isDefined) {
+      if (resolvedMethodOption.get.size == 1) {
+        dstGraph.addEdgeInOriginal(call, resolvedMethodOption.get(0), EdgeTypes.CALL)
+      } else {
+        val withNumParams =
+          resolvedMethodOption.get
+            .map(_.asInstanceOf[Method])
+            .map { method =>
+              method.parameter.size -> method
+            }
```

Comment:
Dont do this for each call site. Do it only once per method during creation of `methodNameToNodes` and store a sequence of tuples instead of just a sequence of methods.

Reply:
yes, much better idea, thanks!